internal/pnp/pnp.go

```diff
@@ -0,0 +1,142 @@
+package pnp
+
+import (
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"testing"
+)
+
+var (
+	isPnpApiInitialized atomic.Uint32
+	cachedPnpApi        *PnpApi
+	pnpMu               sync.Mutex
+	// testPnpCache stores per-goroutine PnP APIs for test isolation
+	// Key is goroutine ID (as int)
+	testPnpCache sync.Map // map[int]*PnpApi
+)
+
+// getGoroutineID returns the current goroutine ID
+// It is usually not recommended to work with goroutine IDs, but it is the most non-intrusive way to setup a parallel testing environment for PnP API
+func getGoroutineID() int {
+	var buf [64]byte
+	n := runtime.Stack(buf[:], false)
+	idField := strings.Fields(strings.TrimPrefix(string(buf[:n]), "goroutine "))[0]
+	id, _ := strconv.Atoi(idField)
+	return id
+}
+
```

Comment:
All of this global / goroutine local storage is something that won't work. If we are storing something, it needs to be attached to a Program, Project, a Host, etc, not global. Parsing out the goroutine ID is definitely a bad idea, especially given our LSP can handle multiple requests at the same time from multiple goroutines and so on.